### PR TITLE
Fix security flaw that attacker can change password of any user.

### PR DESCRIPTION
**Pull request by aihturie**
Attacker can try to send a changing password request along with the uname of another existing user.
BasicAuthenticator.UpdateRecord did not check the existing of the uname.
MySQL adapter also blindly update the record of another user to the new password.

### CODE REVIEW

File: server/auth/basic/auth_basic.go [hunk 1]
```diff
@@ -167,6 +167,11 @@ func (a *authenticator) UpdateRecord(rec *auth.Rec, secret []byte) (*auth.Rec, e
 		uname = login
 	} else if err = a.checkLoginPolicy(uname); err != nil {
 		return nil, err
+	} else if uid, _, _, _, err := store.Users.GetAuthUniqueRecord(a.name, uname); err != nil {
+		return nil, err
+	} else if !uid.IsZero() {
+		// .User is changing unique name, check for duplicate unique name
```
Comment (aforge) [inline server/auth/basic/auth_basic.go:173]: This comment is a little confusing.
Maybe rephrase as `The (new) user name already exists. Report an error.`?

File: server/db/mysql/adapter.go [hunk 2]
```diff
@@ -719,8 +719,8 @@ func (a *adapter) AuthUpdRecord(uid t.Uid, scheme, unique string, authLvl auth.L
 		exp = &expires
 	}
 
-	_, err := a.db.Exec("UPDATE auth SET uname=?,authLvl=?,secret=?,expires=? WHERE uname=?",
-		unique, authLvl, secret, exp, unique)
+	_, err := a.db.Exec("UPDATE auth SET uname=?,authLvl=?,secret=?,expires=? WHERE userid=? AND scheme=?",
```
Comment (aforge) [inline server/db/mysql/adapter.go:722]: It looks like mongo and rethinkdb still have the same problem.
Would it be possible for you to update those as well?

Comment (aihturie) [reply server/db/mysql/adapter.go:722]: it's weird mongodb and rethinkdb did not have this problem.

File: server/db/mysql/adapter.go [hunk 3]
```diff
@@ -719,8 +719,8 @@ func (a *adapter) AuthUpdRecord(uid t.Uid, scheme, unique string, authLvl auth.L
...
+		unique, authLvl, secret, exp, store.DecodeUid(uid), scheme)
 	if isDupe(err) {
```
Comment (or-else) [inline server/db/mysql/adapter.go:724]: This check is redundant.

Comment (or-else) [reply server/db/mysql/adapter.go:724]: Actually, maybe not. Let's leave it there.

**Comment by or-else** [comment]
Thanks!
